### PR TITLE
Optimize primitive collect methods for primitive Bags

### PR DESCRIPTION
**Pull request by LuizFrra**
https://github.com/eclipse/eclipse-collections/issues/420

### CODE REVIEW

**Comment by mohrezaei** [comment]
Since this is a generic `collect` method, should we also optimize for when the target is a `Set`?

**Comment by LuizFrra** [comment]
I can add a check for when target is a Set add only once, What you think about ?

**Comment by mohrezaei** [comment]
@LuizFrra Sounds good.

**Comment by LuizFrra** [comment]
Done, I also add the optimization for the others primitive collect methods.

File: eclipse-collections-code-generator/src/main/resources/api/bag/immutablePrimitiveBag.stg [hunk 1]
```diff
@@ -14,11 +14,16 @@ body(type, name) ::= <<
 
 package org.eclipse.collections.api.bag.primitive;
 
+import java.util.Collection;
+import java.util.Set;
```
Comment (mohrezaei) [inline eclipse-collections-code-generator/src/main/resources/api/bag/immutablePrimitiveBag.stg:18]: I was thinking about the primitive `Set`, not `java.util.Set` (primitive sets don't implement that).

**Review by mohrezaei — CHANGES_REQUESTED** [review]

**Comment by mohrezaei** [comment]
Woops, sorry, I didn't realize the signature contained a to-object function. 

I don't think this optimization is valid because of the function, which may have state (and do things like counting).

**Comment by LuizFrra** [comment]
I can't figure out why the optimization isnt't valid. should I close de PR or just remove the optimization for the ImmutablePrimitiveBag ? Could You elaborate more please ?

Sorry, English isn't my main language.

**Comment by mohrezaei** [comment]
@LuizFrra the issue is subtle. I commented on the original issue #420 with an example function that will break with the optimization.

**Pull request closed without merging**